certbot/error_handler.py

```diff
@@ -34,29 +42,40 @@ class works best as a context manager. For example:
     if a signal is encountered, cleanup_func is called followed by the
     previously registered signal handler.
 
-    Every registered function is attempted to be run to completion
-    exactly once. If a registered function raises an exception, it is
-    logged and the next function is called. If a (different) handled
-    signal occurs while calling a registered function, it is attempted
-    to be called again by the next signal handler.
+    Each registered cleanup function is called exactly once. If a registered
+    function raises an exception, it is logged and the next function is called.
+    Signals received while the registered functions are executing are
+    deferred until they finish.
 
     """
     def __init__(self, func=None, *args, **kwargs):
+        self.body_executed = False
         self.funcs = []
         self.prev_handlers = {}
+        self.received_signals = []
         if func is not None:
             self.register(func, *args, **kwargs)
 
     def __enter__(self):
         self.set_signal_handlers()
```

Comment:
nit: While not necessary based on our current usage, we never reset `body_executed` meaning this class can only be used once. We can easily fix the problem by setting also the `body_executed` in the beginning of `__enter__` or the of `__exit__`. I don't really care, but I prefer the beginning of `__enter__`.